Store constraints and example inputs in the graph module as metadata in export

guangy10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99961

Metadata to store in the GraphModule:
 - input shape constraints
 - example inputs
 - other inline constraints


The saved constraints (in mem) will be used directly after export to convert constraints to runtime assertion which is a separate pass after export. 
The requirement of saved constraints:
  1. Be able to locate where the constraints is from
  2. Should not break the exported graph module serialization.


Examples of saved constraints
```
input_shape_constraints:
  {'t_id': 140266058179792, 'dim': 0, 'min': 6, 'max': oo}
  {'t_id': 140266058179792, 'dim': 0, 'min': 2, 'max': 10}
  
inline_constraints:
  i1: ValueRanges(lower=2, upper=5)
```


cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

guangy10 committed:
Store constraints and example inputs in the graph module as metadata in export

[ghstack-poisoned]

guangy10 committed:
Update on "Store constraints and example inputs in the graph module as metadata in export"


Metadata to store in the GraphModule:
 - input shape constraints
 - example inputs
 - other range constraints which may not be preserved after `make_fx`


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

guangy10 committed:
Update on "Store constraints and example inputs in the graph module as metadata in export"


Metadata to store in the GraphModule:
 - input shape constraints
 - example inputs
 - other range constraints which may not be preserved after `make_fx`


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

guangy10 commented on torch/_dynamo/eval_frame.py:
```diff
@@ -884,6 +887,10 @@ def graph_with_interpreter(*args):
 
+    new_graph.meta["example_inputs"] = example_inputs
+    new_graph.meta["input_shape_constraints"] = constraints
+    new_graph.meta["node_range_constraints"] = node_range_constraints
```
It looks like what gets stored here is a dup of input constraints
```
input_shape_constraints:  [
    Constraint(w_tensor=<weakref at 0x7f2ac6eef3d0; to 'Tensor' at 0x7f2ad64c7f60>, t_id=139821960691552, dim=0, constraint_range=StrictMinMaxConstraint(warn_only=False, vr=ValueRanges(lower=2, upper=oo))), 
    Constraint(w_tensor=<weakref at 0x7f2ac6dae890; to 'Tensor' at 0x7f2ac7fb5f30>, t_id=139821720493872, dim=0, constraint_range=StrictMinMaxConstraint(warn_only=False, vr=ValueRanges(lower=2, upper=oo))), 
    Constraint(w_tensor=<weakref at 0x7f2ac6daea70; to 'Tensor' at 0x7f2ac7fb5030>, t_id=139821720490032, dim=0, constraint_range=StrictMinMaxConstraint(warn_only=False, vr=ValueRanges(lower=2, upper=oo)))
]
node_range_constraints:  {
    s0: ValueRanges(lower=2, upper=9223372036854775806), 
    s1: ValueRanges(lower=2, upper=9223372036854775806), 
    s2: ValueRanges(lower=2, upper=9223372036854775806)
}
```

When I check the id of underlying `ValueRanges` object gets stored, they are different. I assume they should be identical

avikchaudhuri replied:
Why store `var_to_range`, then? I'm not surprised that the ids are different, because we do `&` on value ranges everywhere. Also, in general input constraints will subsume this information.

tugsbayasgalan replied:
@guangy10 could you try saving/loading the graph module produced from dynamo after your change? I remember there are some complications when we use weakref inside a graph module for other users who call save/load on torchdynamo.export graph. If it doesnt work, you can just delete the weakref. I will use tensor_id and example inputs to do the matching.

tugsbayasgalan replied:
And you can delete the constraints on placeholder variables as they will be subsumed by the input constraints

guangy10 replied:
> could you try saving/loading the graph module produced from dynamo after your change? I remember there are some complications when we use weakref inside a graph module for other users who call save/load on torchdynamo.export graph. If it doesnt work, you can just delete the weakref. I will use tensor_id and example inputs to do the matching.

@tugsbayasgalan As you said, save/load doesn't work with weakref inside graph module. 
```
TypeError: cannot pickle 'weakref.ReferenceType' object
```
So I will delete each weakref prior to saving the constaints as metadata

guangy10 committed:
Update on "Store constraints and example inputs in the graph module as metadata in export"


Metadata to store in the GraphModule:
 - input shape constraints
 - example inputs
 - other range constraints which may not be preserved after `make_fx`


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

guangy10 commented on torch/_dynamo/eval_frame.py:
```diff
@@ -884,6 +888,16 @@ def graph_with_interpreter(*args):
     ).transform()
 
+    new_graph.meta["example_inputs"] = example_inputs
+    new_graph.meta["input_shape_constraints"] = constraints
```
It looks like @ezyang would prefer keep `w_tensor` but get rid of `t_id` in the long run (There is 
a TODO). I verified that `weakref` can not be serialized however (doesn't work with `torch.save/load`). IIRC the exported Aten graph should be serializable right. Do we want to fix serialization of `weakref` or we want to keep `t_id` instead?

```
class Constraint:
    """
    This represents constraints on input tensor dimensions, e.g., requiring
    them to be fully polymorphic or within some range.  Don't create this
    class directly; instead, use :func:`torch._export.dynamic_dim`.
    """
    w_tensor: weakref.ReferenceType[torch.Tensor]
    # TODO: We don't need t_id; we can get it off of w_tensor
    t_id: int
    ```

guangy10 replied:
Tried a workaround by setting `w_tensor = None` which still still not make `Constraint` serializable. @tugsbayasgalan if you need it to be saved/loaded in your PR, maybe we can store `Constraint` as key-value pairs w/o `w_tensor` field.

tugsbayasgalan replied:
I don't need it to be serializable. But there are some other people who rely on the fact that torchdynamo.export result is serializable. If we make the graph module not serializable due to w_tensor, we will break their flow. I think we can just do key/value pair that somehow references `t_id` in some form.

tugsbayasgalan replied:
More concretely, I think the format can be in this form:
```
ConstraintSpec = namedtuple(["constraint_dimension", "min", "max"])

# key is tensor id and value is list of constraints for that tensor
input_shape_constraints: Dict[int, List[ConstraintSpec]]
```

avikchaudhuri replied:
If you want to avoid tensor ids, another way to index the constraints is by source. I'm doing something related here: https://github.com/pytorch/pytorch/pull/99993 by adding a tensor id to source map in `tx.output.tracked_fakes_id_to_source` which is populated when fake tensors are created.

Since the sources are related to placeholders, I think this would be a stable representation.

ezyang replied:
Serializing t_id is a bad idea anyway, because it's literally the pointer address from the Python process doing the serialization. Want deterministic export artifacts? Kiss them good bye!

guangy10 replied:
Will save input constraints as it since the graph module serialization is broken already. A proper fix to make it serializable is needed

guangy10 commented on torch/_dynamo/eval_frame.py:
```diff
@@ -884,6 +888,16 @@ def graph_with_interpreter(*args):
 
+    new_graph.meta["example_inputs"] = example_inputs
+    new_graph.meta["input_shape_constraints"] = constraints
+    new_graph.meta["inline_constraints"] = {
+        node.meta["val"].node.expr: var_to_range_map[node.meta["val"].node.expr]
+        for node in new_graph.graph.nodes
+        if node.op != "placeholder"
+        and "val" in node.meta
+        and re.match(r"^i\d+$", str(node.meta["val"]))
+    }
```
@tugsbayasgalan FYI, to retrieve the inline constraints, we have to do this anyway, I decide to save it in the graph module so you can use it directly anywhere else.

guangy10 commented on torch/_dynamo/eval_frame.py:
```diff
@@ -884,6 +888,16 @@ def graph_with_interpreter(*args):
+        for node in new_graph.graph.nodes
+        if node.op != "placeholder"
+        and "val" in node.meta
+        and re.match(r"^i\d+$", str(node.meta["val"]))
```
@ezyang qq about unbacked symints: Are they always named like "i0", "i1", "i2", etc.?

ezyang replied:
Yes but you're not supposed to rely on this...

guangy10 replied:
OK, then what is the recommended way to find constraint from unbacked symint?

guangy10 commented:
OK. It turns out that the exported graph is not serializable w/o any changes in this PR. Even if we ensure the stored metadata are serializable (w/o having `weakref`), it's still not serializable. We should file an issue and address it separately. @tugsbayasgalan @ezyang 

To repro:
```
gm, _ = torch._dynamo.export(…)
buffer = io.BytesIO()
torch.save(gm, buffer)
```

Error like
```
    pickler.dump(obj)
AttributeError: Can't pickle local object 'WeakValueDictionary.__init__.<locals>.remove'
```

guangy10 commented:
OK. Confirmed the root cause is this [line](https://www.internalfb.com/code/fbsource/[8c152b6eb957e911021308b1144a108a6463e95b]/fbcode/caffe2/torch/_dynamo/eval_frame.py?lines=983), fix is in https://github.com/pytorch/pytorch/pull/99461

guangy10 committed:
Update on "Store constraints and example inputs in the graph module as metadata in export"


Metadata to store in the GraphModule:
 - input shape constraints
 - example inputs
 - other inline constraints


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

guangy10 commented on torch/_dynamo/eval_frame.py:
```diff
@@ -884,6 +897,21 @@ def graph_with_interpreter(*args):
     ).transform()
 
+    # Store constraints and inputs as metadata for user passes, e.g. turn constraints to runtime check
+    new_graph.meta["example_inputs"] = example_inputs
+    new_graph.meta["input_shape_constraints"] = (
+        [constraint.serializable_spec for constraint in constraints]
+        if constraints
+        else None
+    )
+    new_graph.meta["inline_constraints"] = {
```
@tugsbayasgalan BTW, the metadata stored in exported graph module won’t be preserved after `torch.store()/load()`. My understanding is that it’s expected as those are not parameters and buffer. If the exported graph gets saved/serialized prior to run the pass turning it into runtime check, they lost it.

tugsbayasgalan replied:
Yep that is fine. Primary reason for store and load to work is to not break other people's flow. I will directly access them, so should be good

guangy10 committed:
Update on "Store constraints and example inputs in the graph module as metadata in export"


Metadata to store in the GraphModule:
 - input shape constraints
 - example inputs
 - other inline constraints


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

guangy10 committed:
Update on "Store constraints and example inputs in the graph module as metadata in export"


Metadata to store in the GraphModule:
 - input shape constraints
 - example inputs
 - other inline constraints


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented:
What are you planning to do with the serialized constraints anyway?

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -671,6 +672,15 @@ def __bool__(self):
+    @property
+    def serializable_spec(self):
+        return {
+            "t_id": self.t_id,
```
For this to be truly useful in the serialized format, you need to know not an opaque id() of the tensor in question, but rather, which positional argument the constraints is being applied to

guangy10 replied:
OK. It looks like there are some confusion about why the constraints need to be serializable and how the serialized constraints will be used.

 1. The reason to save the constraints in a serializable format is just to not break the graph module serialization. The serialized format of the constraint itself doesn't matter. It will not be preserved after save/load.
 2. The saved constraints are used in memory directly after export to convert constraints to runtime assertion which is a separate pass after export.

guangy10 commented:
> What are you planning to do with the serialized constraints anyway?

There will be user passes to turn constraints into runtime assertion after export, so we need to save the constraints for it.

guangy10 commented:
FYI. Updated summary to explain what constraints to save and why. And how the saved ones look like

guangy10 committed:
Update on "Store constraints and example inputs in the graph module as metadata in export"


Metadata to store in the GraphModule:
 - input shape constraints
 - example inputs
 - other inline constraints


The saved constraints (in mem) will be used directly after export to convert constraints to runtime assertion which is a separate pass after export. 
The requirement of saved constraints:
  1. Be able to locate where the constraints is from
  2. Should not break the exported graph module serialization.


Examples of saved constraints
```
input_shape_constraints:
  {'t_id': 140266058179792, 'dim': 0, 'min': 6, 'max': oo}
  {'t_id': 140266058179792, 'dim': 0, 'min': 2, 'max': 10}
  
inline_constraints:
  i1: ValueRanges(lower=2, upper=5)
```


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented:
> > What are you planning to do with the serialized constraints anyway?
> 
> @ezyang There will be user passes to turn constraints into runtime assertion after export, so we need to save the constraints for it.

Constraints seems like the wrong thing to serialize though. Don't you want to serialize the guards? For example, let's say that some inputs are expected to have static size. I don't think you are forcing users to annotate every input with static size in constraints? (Or are you?)

guangy10 commented:
> Constraints seems like the wrong thing to serialize though. Don't you want to serialize the guards? For example, let's say that some inputs are expected to have static size. I don't think you are forcing users to annotate every input with static size in constraints? (Or are you?)

@ezyang Here are my thoughts after discussing with the team.
1. No, we don't need to store info of specialized dims. Those can be figured out by just looking at the example inputs, any dim that is not in the constraints is considered specialized.
2. As today, the inline constraints are not part of guards, so those needs to be stored as metadata anyway. The main motivation of storing all, e.g. example inputs, input constraints, inline constraints, as metadata is just to make it most convenient to fetch those in the pass where they are turned into runtime assertions.
3. To me in the long-term I feel it's better to make the exported module self-contained including the info that holds it sounded, either guards or constraints but maybe not both.

guangy10 commented:
@pytorchbot merge

guangy10 committed:
Update on "Store constraints and example inputs in the graph module as metadata in export"


Metadata to store in the GraphModule:
 - input shape constraints
 - example inputs
 - other inline constraints


The saved constraints (in mem) will be used directly after export to convert constraints to runtime assertion which is a separate pass after export. 
The requirement of saved constraints:
  1. Be able to locate where the constraints is from
  2. Should not break the exported graph module serialization.


Examples of saved constraints
```
input_shape_constraints:
  {'t_id': 140266058179792, 'dim': 0, 'min': 6, 'max': oo}
  {'t_id': 140266058179792, 'dim': 0, 'min': 2, 'max': 10}
  
inline_constraints:
  i1: ValueRanges(lower=2, upper=5)
```


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

guangy10 commented:
> The most logical place for an inline constraint is a torch._assert (courtesy of Tugsuu) inside the graph; it seems awfully circuitous to fail to put them in the graph, and then post facto have to reinsert then inside.

Yeah, discussed with Tugsuu prior to creating this PR. That's what he needs in #98878

guangy10 commented:
@pytorchbot merge

ezyang commented:
What I mean is, if the inline constraint is already part of the graph (which it would be, with working torch._assert), you don't need to *separately* export it.
